src/python/pants/option/optionable.py

```diff
@@ -28,6 +44,130 @@ class Optionable(AbstractClass):
+  def _compute_stable_name(cls):
+    return '{}_{}'.format(cls.__module__, cls.__name__).replace('.', '_')
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return tuple()
+
+  @classmethod
+  def subsystem_dependencies_iter(cls):
+    for dep in cls.subsystem_dependencies():
+      if isinstance(dep, SubsystemDependency):
+        yield dep
+      else:
+        yield SubsystemDependency(dep, GLOBAL_SCOPE)
+
+  def _options(self):
+    return None
+
+  @classmethod
+  def subscope(cls, scope):
+    if cls.options_scope is None or cls.options_scope_category is None:
+      raise OptionsError(
+        '{} must set options_scope and options_scope_category.'.format(cls.__name__))
+    cls.validate_scope_name(cls.options_scope)
+    if scope is None:
+      raise OptionsError('TODO: err msg')
+    if scope == GLOBAL_SCOPE:
+      ret = cls.options_scope
+    else:
+      ret = '{0}.{1}'.format(cls.options_scope, scope)
+    cls.validate_scope_name(ret)
+    return ret
+
+  @classmethod
+  def get_scope_info(cls, subscope=None):
+    """Returns a ScopeInfo instance representing this Optionable's options scope."""
+    if subscope is None:
+      combined_scope = cls.options_scope
+    else:
+      combined_scope = cls.subscope(subscope)
+    return ScopeInfo(combined_scope, cls.options_scope_category, cls)
+
+
+  # FIXME: add subscope arg and make this recursive!
+  @classmethod
+  def known_scope_infos(cls):
+    """Yields ScopeInfo for all known scopes for this task, in no particular order."""
+    # The task's own scope.
+    yield cls.get_scope_info()
+    # The scopes of any task-specific subsystems it uses.
+    for dep in cls.subsystem_dependencies_iter():
+      if not dep.scope == GLOBAL_SCOPE:
+        yield dep.subsystem_cls.get_scope_info(subscope=dep.scope)
+
+  @classmethod
+  def closure_scope_info_strs(cls):
+    q = [(ss, OrderedSet()) for ss in cls.subsystem_dependencies_iter()]
+    known_subsystem_types = set()
+    if cls.options_scope is None:
+      raise Exception('TODO: err msg (cls: {})'.format(cls))
+    known_scopes = set([cls.options_scope])
+
+    while len(q) > 0:
+      (sdep, ss_path) = q.pop()
+      ss = sdep.subsystem_cls
+      if ss in ss_path:
+        cycle = list(ss_path) + [ss]
+        raise cls.CycleException(cycle)
+      if ss not in known_subsystem_types:
+        known_subsystem_types.add(ss)
+        joined_scope = sdep.subsystem_dependency_joined_scope()
+        if joined_scope in known_scopes:
+          raise Exception('TODO: err msg (joined_scope: {})'.format(joined_scope))
+        known_scopes.add(joined_scope)
+      for dep in ss.subsystem_dependencies_iter():
+        q.append((dep, ss_path | [ss]))
+
+    return known_scopes
+
+  @memoized_property
+  def fingerprint(self):
+    hasher = sha1()
+    hasher.update(self.stable_name())
+    hasher.update(self.implementation_version_str())
+    for scope_info in self.closure_scope_info_strs():
+      hasher.update(scope_info)
```

Comment:
Shouldn't this be hasher.update(self._options_fingerprint(scope_str)) ? Looks like you have it that way in Task.fingerprint, but that should presumably be here, no?